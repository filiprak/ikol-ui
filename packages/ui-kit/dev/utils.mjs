import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
