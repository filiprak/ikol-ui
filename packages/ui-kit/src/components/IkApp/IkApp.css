body {
    position: relative;
    background-color: var(--background-color);
    color: var(--on-background-color-1);
    font-family: "Open Sans", sans-serif;
    letter-spacing: 0;
    font-weight: 600;
    font-size: 13px;
}

::selection {
    background-color: var(--primary-color) !important;
    color: var(--on-primary-color) !important;
}

::-moz-selection {
    background-color: var(--primary-color) !important;
    color: var(--on-primary-color) !important;
}

::placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

::-moz-placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

::-webkit-input-placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

:-ms-input-placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

::-ms-input-placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

/* number input arrows */
input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
}

input[type=number] {
    -moz-appearance: textfield;
}

.ik-app {
    display: grid;
    align-content: start;
    position: relative;
    -webkit-backface-visibility: hidden;
    backface-visibility: hidden;
    min-height: 100%;
    max-width: 100%;
}

.ik-app--fill {
    width: 100%;
    height: 100%;
    align-content: normal;
}

.ik-app-bar {
    position: fixed;
    top: 0;
    left: 0;
    flex: 0 0 auto;
    display: flex;
    align-items: center;
    width: 100%;
    max-width: 100%;
    z-index: var(--z-index-app-bar);
    background-color: var(--background-color);
    border-bottom: 1px solid var(--surface-darken-color);
}

.ik-app-bar__back,
.ik-app-bar__burger {
    padding-left: 15px;
    font-size: 24px;
}

.ik-app-bar__title {
    max-width: 50%;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    font-size: 18px;
    padding-left: 15px;
    min-width: 0;
}

.ik-app--mobile .ik-app-bar__title {
    max-width: 100%;
}

.ik-app-bar__content {
    flex: 1 1 auto;
    padding: 0 25px;
    min-width: 0;
}

.ik-app-content {
    display: grid;
    min-height: 0;
}

input::-webkit-contacts-auto-fill-button {
    visibility: hidden;
    display: none !important;
    pointer-events: none;
    position: absolute;
    right: 0;
}
