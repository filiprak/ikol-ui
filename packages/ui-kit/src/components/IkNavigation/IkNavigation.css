.ik-navigation {
    position: fixed;
    left: 0;
    top: 0;
    z-index: var(--z-index-app-nav);
    display: flex;
    flex-direction: column;

    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
    transition: width 0.2s ease-in-out, transform 0.2s ease-in-out;
}

.ik-theme--dark .ik-navigation:not(.ik-navigation--secondary) {
    background-color: var(--surface-color);
    color: var(--on-surface-color-1);
}

.ik-navigation--secondary {
    position: relative;
    z-index: initial;
    background: none;
    width: 250px;
    color: var(--on-background-color-3);
    transition: none;
    max-height: 100%;
}

.ik-navigation__content {
    display: flex;
    flex-direction: column;
    flex: 1 1 auto;
    min-height: 0;
}

.ik-navigation__backdrop {
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    background-color: #000;
    opacity: .5;
    z-index: 3;
}

.ik-theme--dark .ik-navigation__backdrop {
    background-color: #4e4e4e;
}

.ik-navigation--secondary .ik-navigation__content {
    overflow: auto;
}

.ik-navigation--mobile.ik-navigation--secondary .ik-navigation__content {
    flex-direction: row;
    overflow: auto;
}

.ik-navigation__top,
.ik-navigation__bottom {
    flex: 0 0 auto;
}

.ik-navigation-item__link {
    position: relative;
    display: flex;
    justify-content: flex-start;
    align-items: center;
    flex-direction: row;
    padding: 0 20px 15px;
    outline: none;
    text-decoration: none !important;
}

.ik-navigation--mini .ik-navigation-item__link {
    text-align: center;
    padding: 0 10px 15px;
    justify-content: center;
}

.ik-navigation-item__link,
.ik-navigation-item__link:active,
.ik-navigation-item__link:focus,
.ik-navigation-item__link:visited,
.ik-navigation-item__link:hover {
    outline-width: 0;
    color: var(--on-surface-color-3);
}

.ik-navigation--secondary .ik-navigation-item__link {
    flex-direction: row;
    color: inherit;
    height: 48px;
    padding: 0 16px 0 15px;
    margin: 0;
}

.ik-navigation-item--clickable {
    cursor: pointer;
    user-select: none;
}

.ik-navigation-item__icon {
    flex: 0 0 36px;
    width: 36px;
    height: 36px;
    font-size: 15px;
    display: flex;
    justify-content: center;
    align-items: center;
    text-align: center;
    transition: opacity 0.1s;
    -webkit-transition: opacity 0.1s;
}

.ik-navigation-item--circle-icon > .ik-navigation-item__link > .ik-navigation-item__icon {
    border-radius: 50%;
    background-color: rgba(255, 255, 255, 0.08);
}

.ik-navigation-item--active > .ik-navigation-item__link,
.ik-navigation-item--active > .ik-navigation-item__link:active,
.ik-navigation-item--active > .ik-navigation-item__link:hover,
.ik-navigation-item--active > .ik-navigation-item__link:focus,
.ik-navigation-item--active > .ik-navigation-item__link:visited {
    color: var(--on-primary-color);
}

.ik-navigation:not(.ik-navigation--secondary) .ik-navigation-item--active > .ik-navigation-item__link > .ik-navigation-item__icon {
    background-color: var(--primary-color);
}

.ik-navigation--secondary .ik-navigation-item--active > .ik-navigation-item__link,
.ik-navigation--secondary .ik-navigation-item--clickable > .ik-navigation-item__link:hover {
    background-color: var(--primary-color-o1);
    color: var(--primary-color);
}

.ik-navigation--mobile.ik-navigation--secondary .ik-navigation-item--active {
    border-bottom: 3px solid var(--primary-color);
}

.ik-navigation--mobile.ik-navigation--secondary .ik-navigation-item--active,
.ik-navigation--mobile.ik-navigation--secondary .ik-navigation-item--clickable:hover {
    background-color: var(--background-color) !important;
}

.ik-navigation--mini .ik-navigation-item__content {
    font-size: 10px;
    padding: 5px 0;
}

.ik-navigation-item__content {
    padding-left: 10px;
    flex-grow: 1;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
}

.ik-navigation--mobile.ik-navigation--secondary .ik-navigation-item__content {
    padding: 0;
}

.ik-navigation-item--clickable:not(.ik-navigation-item--active):hover .ik-navigation-item__icon {
    opacity: 0.7;
}

.ik-navigation-item__popover {
    background: var(--surface-color);
    color: var(--on-surface-color-1);
    box-shadow: 0 0 20px 8px rgba(31, 44, 56, 0.15);
    border-radius: 16px;
    overflow: hidden;
}

.ik-navigation:not(.ik-navigation--mini) .ik-navigation-item__children {
    padding: 10px 0;
    margin-bottom: 15px;
    background-color: rgba(0, 0, 0, 0.2);
}

.ik-navigation:not(.ik-navigation--mini) .ik-navigation-item__children .ik-divider__line {
    border-color: rgba(255, 255, 255, 0.1);
}

.ik-navigation-item__popover .ik-navigation-item__children .ik-divider {
    padding: 0 15px;
}

.ik-navigation-item__popover .ik-navigation-item__children .ik-divider__line {
    border-color: var(--border-color-2);
}

.ik-navigation-item__children .ik-navigation-item__link {
    display: flex;
    flex-direction: row;
    height: 42px;
    padding: 0 16px 0 66px;
    margin: 0;
    text-align: left;
}

.ik-navigation-item__popover .ik-navigation-item__link {
    padding: 0 16px 0 15px;
}

.ik-navigation-item__children .ik-navigation-item__top > .ik-navigation-item__link {
    color: var(--on-surface-color-1);
}

.ik-navigation-item__children .ik-navigation-item__content {
    font-size: 13px;
    padding: 0;
}

.ik-navigation-item__children .ik-navigation-item:hover > .ik-navigation-item__link,
.ik-navigation-item__children .ik-navigation-item--active > .ik-navigation-item__link {
    background-color: var(--primary-color-o1);
    color: var(--primary-color);
}

.ik-navigation-item__children .ik-navigation-item__icon {
    background: none !important;
    width: 28px;
    height: 28px;
    justify-content: flex-start;
}

.ik-navigation-item__arrow {
    min-width: 0;
}
