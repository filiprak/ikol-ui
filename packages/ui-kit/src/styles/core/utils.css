/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
