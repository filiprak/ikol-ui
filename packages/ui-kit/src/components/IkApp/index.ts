import IkApp from './IkApp.vue';
import IkAppContent from './IkAppContent.vue';
import IkAppBar from './IkAppBar.vue';

export interface IkAppContent extends InstanceType<typeof IkAppContent> { };
export interface IkAppBar extends InstanceType<typeof IkAppBar> { };
export interface IkApp extends InstanceType<typeof IkApp> { };

export { IkApp };
export { IkAppContent };
export { IkAppBar };
