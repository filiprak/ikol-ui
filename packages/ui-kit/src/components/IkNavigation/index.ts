import IkNavigation from './IkNavigation.vue';
import IkNavigationItem from './IkNavigationItem.vue';

export interface IkNavigation extends InstanceType<typeof IkNavigation> { };
export interface IkNavigationItem extends InstanceType<typeof IkNavigationItem> { };

export { IkNavigation };
export { IkNavigationItem };
