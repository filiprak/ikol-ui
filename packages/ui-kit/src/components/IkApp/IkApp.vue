<script lang="ts">
import '@ui/styles';
import './IkApp.css';
import { defineComponent, h, unref } from 'vue';
import { toCssUnits } from '@ui/utils/helpers';

import type { IkAppBar } from '@ui/components/IkApp';
import type { IkNavigation } from '@ui/components/IkNavigation';
import { IkModalRenderer } from '@ui/components/IkModal';
import { useDevice } from '@ui/composables/device';

export default defineComponent({
    provide() {
        return {
            app: this,
        };
    },
    props: {
        loading: {
            type: Boolean,
            default: false,
        },
        fill: {
            type: Boolean,
            default: false,
        },
    },
    data() {
        return {
            nav: null as IkNavigation | null,
            bar: null as IkAppBar | null,
        };
    },
    computed: {
        styles() {
            if (!unref(this.device.mobile)) {
                return {
                    paddingLeft: toCssUnits(this.nav_width),
                    paddingTop: toCssUnits(this.bar_height),
                };
            } else {
                return {
                    paddingTop: toCssUnits(this.bar_height),
                };
            }
        },
        nav_width() {
            if (this.nav) {
                return this.nav.mini_width;
            } else {
                return 0;
            }
        },
        bar_height() {
            if (this.bar) {
                return this.bar.height;
            } else {
                return 0;
            }
        },
    },
    methods: {
        registerBar(bar: IkAppBar) {
            this.bar = bar || null;
        },
        registerNav(nav: IkNavigation) {
            this.nav = nav || null;
        },
        unregisterBar() {
            this.bar = null;
        },
        unregisterNav() {
            this.nav = null;
        },
    },
    render() {
        let children = this.$slots.default?.() || [];

        if (this.loading) {
            children = [];
        }

        children.push(h(IkModalRenderer));

        return h('div', {
            class: {
                'ik-app': true,
                'ik-app--fill': this.fill,
                'ik-app--mobile': unref(this.device.mobile),
            },
            style: this.styles,
        }, children);
    },
    setup() { return { device: useDevice() }; },
});
</script>
