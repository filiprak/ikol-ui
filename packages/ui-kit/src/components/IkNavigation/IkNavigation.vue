<script lang="ts">
import '@ui/styles';
import './IkNavigation.css';
import type { PropType } from 'vue';
import { defineComponent, h, unref, inject } from 'vue';

import type { IkApp } from '@ui/components/IkApp';
import { toCssUnits } from '@ui/utils/helpers';
import { IkScrollArea } from '@ui/components/IkScrollArea';
import { useDevice } from '@ui/composables/device';

export default defineComponent({
    provide() {
        return { navigation: this };
    },
    props: {
        open: {
            type: Boolean,
            default: false,
        },
        mini: {
            type: Boolean,
            default: false,
        },
        tag: {
            type: String,
            default: 'nav',
        },
        width: {
            type: Number,
            default: 256,
        },
        mini_width: {
            type: Number,
            default: 72,
        },
        exclude_close: {
            type: Array as PropType<HTMLElement[]>,
            default() {
                return [];
            },
        },
        secondary: {
            type: Boolean,
            default: false,
        },
    },
    setup() {
        return { app: inject<IkApp>('app'), device: useDevice() };
    },
    data() {
        return {
            backdrop_el: null as HTMLElement | null,
        };
    },
    computed: {
        root_class() {
            return {
                'ik-navigation': true,
                'ik-navigation--mini': this.mini,
                'ik-navigation--secondary': this.secondary,
                'ik-navigation--mobile': unref(this.device.mobile),
            };
        },
        styles() {
            if (!this.secondary) {
                return {
                    height: '100vh',
                    top: 0,
                    maxHeight: '100%',
                    transform: 'translateX(' + (this.open ? '0%' : '-100%') + ')',
                    width: toCssUnits(this.computed_width),
                };
            } else {
                return {
                    width: unref(this.device.mobile) ? '100%' : null,
                };
            }
        },
        computed_width() {
            if (this.mini) {
                return this.mini_width;
            } else {
                return this.width;
            }
        },
    },
    watch: {
        open: {
            immediate: true,
            handler(value) {
                if (unref(this.device.mobile) && !this.secondary) {
                    if (value) {
                        window.addEventListener('mousedown', this._onWindowMouseDown, true);
                    } else {
                        window.removeEventListener('mousedown', this._onWindowMouseDown, true);
                    }
                    this._toggleBackdrop(value);
                }
            },
        },
        mini: {
            immediate: true,
            handler(mini) {
                if (!unref(this.device.mobile) && !this.secondary) {
                    if (!mini) {
                        window.addEventListener('mousedown', this._onWindowMouseDown, true);
                    } else {
                        window.removeEventListener('mousedown', this._onWindowMouseDown, true);
                    }
                }
            },
        },
        'device.mobile.value'(mobile) {
            this.$emit('update:open', !mobile);
            this.$emit('update:mini', !mobile);

            if (!mobile) {
                this._toggleBackdrop(false);
            }
        },
    },
    created() {
        if (!this.secondary) {
            this.app && this.app.registerNav(this);
        }
    },
    mounted() {
        if (!this.secondary) {
            this._initBackdrop();
        }
    },
    beforeUnmount() {
        if (!this.secondary) {
            this.app && this.app.unregisterNav();
            window.removeEventListener('mousedown', this._onWindowMouseDown, true);
        }
        this._destroyBackdrop();
    },
    methods: {
        toggleMini(mini?: boolean) {
            this.$emit('update:mini', mini !== undefined ? !!mini : !this.mini);
        },
        toggle(value?: boolean) {
            if (!this.secondary) {
                if (unref(this.device.mobile)) {
                    this.$emit('update:open', value !== undefined ? !!value : !this.open);
                } else {
                    this.$emit('update:mini', value !== undefined ? !value : !this.mini);
                }
            }
        },
        _onWindowMouseDown(event: MouseEvent) {
            let should_close = this.$el !== event.target && !this.$el.contains(event.target as HTMLElement);

            // check if element not excluded
            if (should_close) {
                for (let i = 0; i < this.exclude_close.length; i++) {
                    if (this.exclude_close[i].contains(event.target as HTMLElement)) {
                        should_close = false;
                        break;
                    }
                }
            }

            if (should_close) {
                this.toggle(false);
                event.preventDefault();
                event.stopPropagation();
            }
        },
        _initBackdrop() {
            const backdrop_el = document.createElement('div');

            backdrop_el.classList.add('ik-navigation__backdrop');
            backdrop_el.style.display = 'none';

            this.backdrop_el = backdrop_el;

            document.body.appendChild(backdrop_el);
        },
        _toggleBackdrop(visible?: boolean) {
            if (this.backdrop_el) {
                this.backdrop_el.style.display = visible ? '' : 'none';
            }
        },
        _destroyBackdrop() {
            if (this.backdrop_el) {
                this.backdrop_el.remove();
                this.backdrop_el = null;
            }
        },
    },
    render() {
        return h(this.tag, {
            class: this.root_class,
            style: this.styles,
        }, {
            default: () => {
                const content = h('div', { class: 'ik-navigation__content' }, [
                    this.secondary && this.device.mobile.value ? this.$slots.default?.() : h(IkScrollArea, {}, { default: () => this.$slots.default?.() }),
                ]);

                const top = h('div', { class: 'ik-navigation__top' }, [
                    this.$slots.top?.(),
                ]);
                const bottom = h('div', { class: 'ik-navigation__bottom' }, [
                    this.$slots.bottom?.(),
                ]);

                return [top, content, bottom];
            },
        });
    },
});
</script>
