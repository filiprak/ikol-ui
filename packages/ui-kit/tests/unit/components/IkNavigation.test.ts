import type { RenderOptions } from '@ui/tests/tools/utils';
import { renderShallow } from '@ui/tests/tools/utils';
import { IkApp } from '@ui/components/IkApp';
import { IkNavigation } from '@ui/components/IkNavigation';

describe('IkNavigation', () => {
    const getInstance = (options: RenderOptions<typeof IkNavigation> = {}) => renderShallow(IkNavigation, {
        global: {
            provide: { app: getAppInstance() },
        },
    }, options);

    const getAppInstance = () => renderShallow(IkApp).vm;

    it('renders with default state', () => {
        const wrapper = getInstance();

        expect(wrapper.classes()).toStrictEqual([
            'ik-navigation',
        ]);

        expect(wrapper.element.tagName).toBe('NAV');
        expect(wrapper.attributes('style')).toBe(
            'height: 100vh; top: 0px; max-height: 100%; transform: translateX(-100%); width: 256px;'
        );
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders on mobile', () => {
        const wrapper = getInstance({
            props: { open: true, mini: false },
            device: { mobile: true },
        });

        expect(wrapper.classes()).toContain('ik-navigation--mobile');
    });

    it('renders opened', () => {
        const wrapper = getInstance({
            props: { open: true },
        });
        expect(wrapper.attributes('style')).toBe(
            'height: 100vh; top: 0px; max-height: 100%; transform: translateX(0%); width: 256px;'
        );
    });

    it('renders mini', () => {
        const wrapper = getInstance({
            props: { open: true, mini: true },
        });
        expect(wrapper.classes()).toContain('ik-navigation--mini');
        expect(wrapper.attributes('style')).toBe(
            'height: 100vh; top: 0px; max-height: 100%; transform: translateX(0%); width: 72px;'
        );
    });

    it('renders with default slot', () => {
        const wrapper = getInstance({
            global: { renderStubDefaultSlot: true },
            props: { open: true },
            slots: { default: '<div>Nav content...</div>' },
        });
        expect(wrapper.text()).toContain('Nav content...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with top slot', () => {
        const wrapper = getInstance({
            props: { open: true },
            slots: { top: '<div>Nav top content...</div>' },
        });
        expect(wrapper.find('.ik-navigation__top').text()).toContain('Nav top content...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with bottom slot', () => {
        const wrapper = getInstance({
            props: { open: true },
            slots: { bottom: '<div>Nav bottom content...</div>' },
        });
        expect(wrapper.find('.ik-navigation__bottom').text()).toContain('Nav bottom content...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('closes full mode on outside click on desktop', () => {
        const wrapper = getInstance({
            props: { open: true, mini: false },
        });

        document.dispatchEvent(new MouseEvent('mousedown'));

        expect(wrapper.emitted('update:mini')).toStrictEqual([
            [true],
        ]);
    });

    it('closes on mobile on outside touch', () => {
        const wrapper = getInstance({
            props: { open: true, mini: false },
            device: { mobile: true },
        });

        document.dispatchEvent(new MouseEvent('mousedown'));

        expect(wrapper.emitted('update:open')).toStrictEqual([
            [false],
        ]);
    });

    it('renders with custom width', () => {
        const wrapper = getInstance({
            props: { open: true, width: 777 },
        });

        document.dispatchEvent(new MouseEvent('mousedown'));

        expect(wrapper.attributes('style')).toBe(
            'height: 100vh; top: 0px; max-height: 100%; transform: translateX(0%); width: 777px;'
        );
    });

    it('renders with custom mini_width', () => {
        const wrapper = getInstance({
            props: { open: true, mini: true, mini_width: 111 },
        });

        document.dispatchEvent(new MouseEvent('mousedown'));

        expect(wrapper.attributes('style')).toBe(
            'height: 100vh; top: 0px; max-height: 100%; transform: translateX(0%); width: 111px;'
        );
    });

    it('works with toggleMini() method', async () => {
        const wrapper = getInstance({
            props: { open: true, mini: true },
        });

        wrapper.vm.toggleMini();

        expect(wrapper.emitted('update:mini')).toStrictEqual([
            [false],
        ]);

        await wrapper.setProps({ mini: false });

        wrapper.vm.toggleMini(true);

        expect(wrapper.emitted('update:mini')).toStrictEqual([
            [false],
            [true],
        ]);
    });

    it('works with toggle() method on desktop', async () => {
        const wrapper = getInstance({
            props: { open: true, mini: true },
        });

        wrapper.vm.toggle();

        expect(wrapper.emitted('update:mini')).toStrictEqual([
            [false],
        ]);

        await wrapper.setProps({ mini: false });

        wrapper.vm.toggle(false);

        expect(wrapper.emitted('update:mini')).toStrictEqual([
            [false],
            [true],
        ]);
    });

    it('works with toggle() method on mobile', () => {
        const wrapper = getInstance({
            props: { open: false, mini: true },
            device: { mobile: true },
        });

        wrapper.vm.toggle();

        expect(wrapper.emitted('update:open')).toStrictEqual([
            [true],
        ]);
    });

    it('renders as secondary navigation', () => {
        const wrapper = getInstance({
            props: { secondary: true },
        });

        expect(wrapper.classes()).toContain('ik-navigation--secondary');
        expect(wrapper.attributes('style')).not.toBeDefined();
        expect(wrapper.find('ik-navigation__toggle').exists()).toBeFalsy();
        expect(wrapper.element).toMatchSnapshot();
    });
});
