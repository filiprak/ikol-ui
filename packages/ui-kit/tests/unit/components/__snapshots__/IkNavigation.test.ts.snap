// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkNavigation renders as secondary navigation 1`] = `
<nav
  class="ik-navigation ik-navigation--secondary"
>
  <div
    class="ik-navigation__top"
  >
    <!---->
  </div>
  <div
    class="ik-navigation__content"
  >
    <ik-scroll-area-stub
      bound_threshold="20"
      disable_auto_anchoring="false"
      hide_loader="false"
      hide_x="false"
      hide_y="false"
      horizontal_arrows="false"
      loading="false"
      pagination="[object Object]"
      reverse="false"
    />
  </div>
  <div
    class="ik-navigation__bottom"
  >
    <!---->
  </div>
</nav>
`;

exports[`IkNavigation renders with bottom slot 1`] = `
<nav
  class="ik-navigation"
  style="height: 100vh; top: 0px; max-height: 100%; transform: translateX(0%); width: 256px;"
>
  <div
    class="ik-navigation__top"
  >
    <!---->
  </div>
  <div
    class="ik-navigation__content"
  >
    <ik-scroll-area-stub
      bound_threshold="20"
      disable_auto_anchoring="false"
      hide_loader="false"
      hide_x="false"
      hide_y="false"
      horizontal_arrows="false"
      loading="false"
      pagination="[object Object]"
      reverse="false"
    />
  </div>
  <div
    class="ik-navigation__bottom"
  >
    
    <div>
      Nav bottom content...
    </div>
    
  </div>
</nav>
`;

exports[`IkNavigation renders with default slot 1`] = `
<nav
  class="ik-navigation"
  style="height: 100vh; top: 0px; max-height: 100%; transform: translateX(0%); width: 256px;"
>
  <div
    class="ik-navigation__top"
  >
    <!---->
  </div>
  <div
    class="ik-navigation__content"
  >
    <ik-scroll-area-stub
      bound_threshold="20"
      disable_auto_anchoring="false"
      hide_loader="false"
      hide_x="false"
      hide_y="false"
      horizontal_arrows="false"
      loading="false"
      pagination="[object Object]"
      reverse="false"
    >
      <div>
        Nav content...
      </div>
    </ik-scroll-area-stub>
  </div>
  <div
    class="ik-navigation__bottom"
  >
    <!---->
  </div>
</nav>
`;

exports[`IkNavigation renders with default state 1`] = `
<nav
  class="ik-navigation"
  style="height: 100vh; top: 0px; max-height: 100%; transform: translateX(-100%); width: 256px;"
>
  <div
    class="ik-navigation__top"
  >
    <!---->
  </div>
  <div
    class="ik-navigation__content"
  >
    <ik-scroll-area-stub
      bound_threshold="20"
      disable_auto_anchoring="false"
      hide_loader="false"
      hide_x="false"
      hide_y="false"
      horizontal_arrows="false"
      loading="false"
      pagination="[object Object]"
      reverse="false"
    />
  </div>
  <div
    class="ik-navigation__bottom"
  >
    <!---->
  </div>
</nav>
`;

exports[`IkNavigation renders with top slot 1`] = `
<nav
  class="ik-navigation"
  style="height: 100vh; top: 0px; max-height: 100%; transform: translateX(0%); width: 256px;"
>
  <div
    class="ik-navigation__top"
  >
    
    <div>
      Nav top content...
    </div>
    
  </div>
  <div
    class="ik-navigation__content"
  >
    <ik-scroll-area-stub
      bound_threshold="20"
      disable_auto_anchoring="false"
      hide_loader="false"
      hide_x="false"
      hide_y="false"
      horizontal_arrows="false"
      loading="false"
      pagination="[object Object]"
      reverse="false"
    />
  </div>
  <div
    class="ik-navigation__bottom"
  >
    <!---->
  </div>
</nav>
`;
